### PR TITLE
Basic authentication for HTTP remote state backend

### PR DESCRIPTION
**Pull request by syskill**
A simple fix to make the HTTP remote state backend pass muster for production use.

### CODE REVIEW

File: state/remote/http.go [hunk 1]
```diff
@@ -44,20 +44,40 @@ func httpFactory(conf map[string]string) (Client, error) {
...
 
 func (c *HTTPClient) Get() (*Payload, error) {
-	resp, err := c.Client.Get(c.URL.String())
+	req, err := http.NewRequest("GET", c.URL.String(), nil)
+	if err != nil {
+		return nil, err
+	}
+
+	// Prepare the request
+	if c.Username != "" {
```
Comment (grubernaut) [inline state/remote/http.go]: What if `username` is set and `password` is not set? Possible to add in extra verification for this case?

File: state/remote/http.go [hunk 2]
```diff
@@ -141,6 +161,9 @@ func (c *HTTPClient) Put(data []byte) error {
 	req.Header.Set("Content-Type", "application/json")
 	req.Header.Set("Content-MD5", b64)
 	req.ContentLength = int64(len(data))
+	if c.Username != "" {
```
Comment (grubernaut) [inline state/remote/http.go]: Same here

File: state/remote/http.go (hunk 1 above)
Comment (syskill) [inline state/remote/http.go]: Unnecesary. The empty string is the zero value, so in this case the HTTP client will just use an empty password, which is totally legit for HTTP basic auth.

File: state/remote/http.go (hunk 2 above)
Comment (syskill) [inline state/remote/http.go]: Same here

**Pull request merged**